firedrake/mesh.py

```diff
@@ -1010,6 +1014,156 @@ def _order_data_by_cell_index(self, column_list, cell_data):
+            raise NotImplementedError("Mesh reordering not implemented for vertex only meshes yet.")
+
+        # BEGIN TODO
+        # dmplex.validate_mesh(plex) # create equivalent for DMSwarm?
+        # END TODO
+
+        swarm.setFromOptions()
+
+        utils._init()
+
+        self._parent_mesh = parentmesh
+        self._swarm = swarm
+        self.name = name
+        self.comm = dup_comm(swarm.comm.tompi4py())
+
+        # A cache of shared function space data on this mesh
+        self._shared_data_cache = defaultdict(dict)
+
+        # Cell subsets for integration over subregions
+        self._subsets = {}
+
+        tdim = 0
+
+        cell = ufl.Cell("vertex")
+        self._ufl_mesh = ufl.Mesh(ufl.VectorElement("DG", cell, 0, dim=cell.topological_dimension()))
+
+        # A set of weakrefs to meshes that are explicitly labelled as being
+        # parallel-compatible for interpolation/projection/supermeshing
+        # To set, do e.g.
+        # target_mesh._parallel_compatible = {weakref.ref(source_mesh)}
+        self._parallel_compatible = None
+
+        def callback(self):
+            """Finish initialisation."""
+            del self._callback
+
+            # Mark OP2 entities and derive the resulting Swarm numbering
+            with timed_region("Mesh: numbering"):
+                dmswarm.mark_entity_classes(self._swarm)
+                self._entity_classes = dmswarm.get_entity_classes(self._swarm).astype(int)
+
+                # Derive a cell numbering from the Swarm numbering
+                entity_dofs = np.zeros(tdim+1, dtype=IntType)
+                entity_dofs[-1] = 1
+
+                self._cell_numbering = self.create_section(entity_dofs)
+                entity_dofs[:] = 0
+                entity_dofs[0] = 1
+                self._vertex_numbering = self.create_section(entity_dofs)
+
+        self._callback = callback
+
+    @property
+    def cell_closure(self):
+        """2D array of ordered cell closures
+
+        Each row contains ordered cell entities for a cell, one row per cell.
+        """
+        swarm = self._swarm
+        tdim = 0
+
+        # Cell numbering and global vertex numbering
+        cell_numbering = self._cell_numbering
+        vertex_numbering = self._vertex_numbering.createGlobalSection(swarm.getPointSF())
+
+        cell = self.ufl_cell()
+        assert tdim == cell.topological_dimension()
+        if cell.is_simplex():
+            import FIAT
+            topology = FIAT.ufc_cell(cell).get_topology()
+            entity_per_cell = np.zeros(len(topology), dtype=IntType)
+            for d, ents in topology.items():
+                entity_per_cell[d] = len(ents)
+
+            return dmswarm.closure_ordering(swarm, vertex_numbering,
+                                            cell_numbering, entity_per_cell)
+        else:
+            raise NotImplementedError("Cell type '%s' not supported." % cell)
+
+    def _facets(self, kind):
+        return None
```

Comment:
You don't need this right?

Reply:
I do because otherwise I default to using the `MeshTopology._facets` method which gives an error because it expects a plex. Really I just need to to return None.

Reply:
Or perhaps attribute error

Reply:
Updated to do the latter.